component/common/loki/wal/watcher.go

```diff
@@ -331,9 +349,24 @@ func (w *Watcher) decodeAndDispatch(b []byte, segmentNum int) (bool, error) {
 	return readData, firstErr
 }
 
-func (w *Watcher) Stop() {
-	// first close the quit channel to order main mainLoop routine to stop
-	close(w.quit)
+// Stop stops the Watcher, draining the WAL until the end of the last segment if drain is true. Since the writer of the WAL
+// is expected to have stopped before the Watcher, no further writes are expected, so segments can be safely consumed.
+//
+// Note if drain is enabled, the caller routine of Stop will block executing the drain procedure.
+func (w *Watcher) Stop(drain bool) {
+	if drain {
+		level.Info(w.logger).Log("msg", "Draining Watcher")
+		w.state.Transition(internal.StateDraining)
+		// wait for drain timeout, or stopping state, in case the Watcher does the transition itself promptly
+		select {
+		case <-time.NewTimer(w.drainTimeout).C:
```

Comment:
Does this mean each wal watcher gets the grace period which could become quite long? IE if you had 4 it would be 2 minutes.

Reply:
Mmm as for this, in theory and how is implemented yes. All watchers are stopped sequentially, and all clients after. Which taking into account what I mentioned [here](https://github.com/grafana/agent/pull/5804#discussion_r1403440247), could lead to `N * (Watcher.DrainTimeout + QueueConfig.DrainTimeout)`, which if one is using many `EndpointConfigs` could be big. 

Technically, the only relation on stopping is that the `(watcher, client)` pairs are stopped in order. wdyt of changing a bit the structure for keeping them in the manager, and doing something like:

1. watchers and client when creating are kept as tuples, like stoppables  = list(<watcher, client>)

```
def stop:
  var wg sync.WaitGroup
  for watch, client in stoppables:
    wg.Add(1)
    go func() {
      watch.Stop()
      client.Stop()
    }

  wg.Wait()
```

in that case we are only spending `Watcher.DrainTimeout + QueueConfig.DrainTimeout` over all endpoint configs.

Reply:
Ok, changed the manager so that all `watcher -> client` pairs are stopped concurrecntly, reducing the graceful stop time by the number of client config.s